src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -346,11 +343,26 @@ private bool IsLegacyAdalCacheEnabled(AuthenticationRequestParameters requestPar
             }
         }
 
-        private DateTimeOffset CalculateSuggestedCacheExpiry()
+        internal /* for testing */ static DateTimeOffset? CalculateSuggestedCacheExpiry(
+            ITokenCacheAccessor accessor, 
+            ICoreLogger logger)
         {
-            IEnumerable<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensWithNoLocks(true);
-            DateTimeOffset cacheExpiry = tokenCacheItems.Max(item => item.ExpiresOn);
-            return cacheExpiry;
+            // If we have refresh tokens in the cache, we cannot suggest expiration
+            // because refresh token expiration is not disclosed to SDKs and RTs are long lived anyway (3 months by default)
+            if (accessor.GetAllRefreshTokens().Count == 0)
+            {
+                IReadOnlyList<MsalAccessTokenCacheItem> tokenCacheItems = accessor.GetAllAccessTokens(optionalPartitionKey: null);
+                if (tokenCacheItems.Count == 0)
+                {
+                    logger.Warning("[CalculateSuggestedCacheExpiry] No access tokens or refresh tokens found in the accessor. Suggesting immediate expiration.");
```

Comment:
@pmaytak - I don't really understand how we can get in this sitatuation, unless there is a race condition in the accessor. The customers complaining are using the AppTokenCache. 

This method is only called when we save a valid MSAL token response, which has a new access token.

Reply:
This only happens when cache serialization is enabled. What about this?
- Thread 1 saves token in the cache.
- Thread 2 clears cache.
- Thread 1 gets to calculating the expiry but the cache is empty.

I updated the NetCore dev app with a multi thread test case [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/ed901db2a22bd860db1050a5bfbda667468247d4/tests/devapps/NetCoreTestApp/Program.cs#L251). So maybe that can be used to repro this, just add in-memory cache serialization.

Reply:
Also our cache is based on eventual consistency, since we don't lock, but before deserialization we clear the cache. So in a multi threaded env, we can get into the above scenario?